### ISSUE TITLE
Schema properties become optional if passed to generic function

### ISSUE THREAD

**Issue by etiennelenhart** [colinhacks/zod#2649]
When passing a schema with a normally matching generic type as a function parameter, object properties seem to become optional. So the following code fails with type error `Type 'undefined' is not assignable to type 'string'.`:
```typescript
// not working
function test1<T extends ZodType<string>>(schema: T) {
  return z.object({ value: schema }) satisfies ZodType<{ value: string }>
                                  // ^ Type 'undefined' is not assignable to type 'string'.
}
```

When passing the schema with a specific type, it is working, but as stated in the documentation, some type information will be lost:
```typescript
// working
function test2(schema: ZodType<string>) {
  return z.object({ value: schema }) satisfies ZodType<{ value: string }>
}
```

I'm not sure whether this is a TypeScript quirk, a bug in Zod or something I'm doing wrong. Any help would be appreciated.

[Reproduction on StackBlitz](https://stackblitz.com/edit/typescript-d7ho4u?file=index.ts&view=editor)
Maybe related: #2198

**Comment by NextThread** [comment]
can i work on this

**Comment by danni** [comment]
I've just run into this when upgrading from Zod 3.18 to 3.22 with the following example:

```ts
type FormValues = {
  obj: {
    key: string
  }
}

const schema: z.Schema<FormValues> = z.object({
  obj: z.object({
    key: z.string().default(''),
  })
})
```

**Comment by JacobWeisenburger** [comment]
@danni,
Is this what you are looking for?

```ts
type FormValuesInput = { obj: { key?: string } }
type FormValuesOutput = { obj: { key: string } }

const schema: z.Schema<FormValuesOutput, z.ZodTypeDef, FormValuesInput> = z.object( {
    obj: z.object( {
        key: z.string().default( '' ),
    } )
} )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by JacobWeisenburger** [comment]
@etiennelenhart,
Is this what you are looking for?

```ts
function test1<T extends z.ZodString> ( schema: T ) {
    return z.object( { value: schema } ) satisfies z.ZodObject<{ value: T }>
}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by danni** [comment]
@JacobWeisenburger so it only works if you supply an input schema?

**Comment by JacobWeisenburger** [comment]
The 1st generic type on z.Schema (the output type), is used as the default value for the 3rd generic type on z.Schema (the input type). As long as they are the same, then you can just specify the 1st one. But in this case they were different because `.default` allows an optional input type, but it returns a required output type.

**Comment by danni** [comment]
Yeah I understand the separate schemas, although I'm wondering here if the type inference makes sense. i.e. one can use the nullish-coalescing operator on a defined value and it's still handled at runtime, even if it's a hypothetical no-op.
